crates/tinymist-query/src/code_action.rs

```diff
@@ -116,4 +120,32 @@ mod tests {
             })
         });
     }
+
+    fn compute_code_action_context(
+        ctx: &mut LocalContext,
+        request_range: &LspRange,
+    ) -> CodeActionContext {
+        let Warned { output, warnings } = typst::compile::<PagedDocument>(&ctx.world);
+        let errors = output.err().unwrap_or_default();
+        let compiler_diagnostics = warnings.iter().chain(errors.iter());
+
+        // Run the linter for additional diagnostics as well.
+        let diagnostics = DiagWorker::new(ctx)
+            .check()
+            .convert_all(compiler_diagnostics)
+            .into_values()
+            .flatten();
+
+        CodeActionContext {
+            diagnostics: diagnostics
+                .filter(|diag| ranges_overlap(&diag.range, request_range))
```

Comment:
The filtering here matches the LSP spec (and VS Code behavior):

> `diagnostics`: An array of diagnostics known on the client side **overlapping the range provided** to the `textDocument/codeAction` request [...]

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#codeActionContext